library/x509write_crt.c

```diff
@@ -510,9 +532,29 @@ int mbedtls_x509write_crt_der(mbedtls_x509write_cert *ctx,
 
     /*
      *  Serial   ::=  INTEGER
+     *
+     * Written data is:
+     * - "ctx->serial_len" bytes for the raw serial buffer
+     *   - if MSb of "serial" is 1, then prepend an extra 0x00 byte
+     * - 1 byte for the length
+     * - 1 byte for the TAG
      */
-    MBEDTLS_ASN1_CHK_ADD(len, mbedtls_asn1_write_mpi(&c, buf,
-                                                     &ctx->serial));
+    MBEDTLS_ASN1_CHK_ADD(len, mbedtls_asn1_write_raw_buffer(&c, buf,
```

Comment:
It would be great to have a test that checks that the written serial has the exact value that we expected. Do we have such test anywhere?

Reply:
Good point, as a general when there's a limit we should check that `len == limit` works and `len == limit + 1` fails.

Reply:
Well, checking the test data it looks like we already have a test case for `MBEDTLS_X509_RFC5280_MAX_SERIAL_LEN` actually: https://github.com/Mbed-TLS/mbedtls/pull/6883/files#diff-c882a142e0059a06b9f052239bff1250881537adaeb385351ce03365fce69e98R145

Reply:
yes, this is also mentioned in a [following comment](https://github.com/Mbed-TLS/mbedtls/pull/6883#discussion_r1088106290).
Perhaps we can remove the `x509_set_serial_check` function and condensate all the tests in `x509_crt_check`?

Reply:
> Good point, as a general when there's a limit we should check that `len == limit` works and `len == limit + 1` fails.

I think that you wanted to write this comment somewhere else :) 
Here I was mentioning that it would be good to have a test that validates that what we write is what we intended to, something like here: https://github.com/Mbed-TLS/mbedtls/blob/169d9e6eb4096cb48aa25651f42b276089841087/tests/suites/test_suite_x509parse.data#L1
So a (possibly) minimal test that would pass the serial number to be written, then would validate that it indeed was written the way we wanted.
Do you think that this is an overkill?

Reply:
I understand the question, but as I was mentioning [here](https://github.com/Mbed-TLS/mbedtls/pull/6883#issuecomment-1405345328) there are already tests which verify that we can get the very same CRT as `openssl` does for the same serial. Isn't this enough (or perhaps better) as guarantee that we're writing the serial in a proper way?

Reply:
Ah, right, I missed that comment. That's good, thanks!